pkgs/pkgs-lib/tests/formats.nix

```diff
@@ -469,4 +469,29 @@ in runBuildTests {
     '';
   };
 
+  xmlGenerate = shouldPass {
+    format = formats.xml { };
+    input = {
+      root = {
+        "@id" = "123";
+        "@class" = "example";
+        child1 = {
+          "@name" = "child1Name";
+          "#text" = "text node";
+        };
+        child2 = {
+          grandchild = "This is a grandchild text node.";
+        };
+      };
+    };
+    expected = ''
+      <?xml version="1.0" encoding="utf-8"?>
+      <root class="example" id="123">
+        <child1 name="child1Name">text node</child1>
+        <child2>
+          <grandchild>This is a grandchild text node.</grandchild>
+        </child2>
+      </root>
+    '';
+  };
```

Comment:
How would namespaces look?

Does definition merging work as expected?

Reply:
it follows: https://www.xml.com/pub/a/2006/05/31/converting-between-xml-and-json.html:

> doesn't take into consideration the following:
> 
> - XML declaration
> - processing instructions
> - explicit handling of namespace declarations
> - XML comments

i found: https://github.com/martinblech/xmltodict/blob/master/README.md#namespace-support

Reply:
namespaces are "mainly problematic" for parsing the other way around. it can be achieved via:

```
{
  "RSS": {
    "@xmlns:jwplayer": "http://support.jwplayer.com/customer/portal/articles/1403635-media-format-reference#feeds",
    "@version": "2.0",
    "Channel": {
...
```
to create:
```
<?xml version="1.0" encoding="utf-16"?>
<RSS xmlns:jwplayer="http://support.jwplayer.com/customer/portal/articles/1403635-media-format-reference#feeds" version="2.0">
  <Channel>
```

Reply:
Ok, makes sense.

Handling foreign input while programming with hardcoded prefixes is basically wrong; an XML document should behave exactly the same under prefix renaming, but a lot of software requires certain prefixes, which is - in principle - wrong.
However, we will have to cater to such applications, and users who are kinda ok with such requirements, so that's just that.

Treating any node as _URI + tag_ instead of _prefix + tag_ makes any XML processing more robust (for an option type, the processing means implementing a `merge` operation for the XML representation), so that may be worth considering, but it seems that we'd end up writing a small library for that purpose; a bit of a research project.

So for now I think we only need to point out in the docs that the type deals with the concrete syntax of the XML, and all responsibility is on the user when it comes to the correctness of any use of namespace features.

Reply:
since this will not be the "perfect/best" xml-exporter for nix, I would suggest to get started, but name it "specially" (not `xml` as we both agreed above.).

Surely there are a lot of things to improve, but in the end we should enable other users to enable new NixOS-options. Migrating this to other formats seems tedious (manual work), but doable.

Documentation is currently missing completely, because I do not know where to put it. Any suggestions where to put it?